packages/web-frontend/src/containers/MapOverlayBar/MapOverlayBar.js

```diff
@@ -43,28 +44,41 @@ const pinnedOverlayCodeReducer = (currentOverlayCode, newOverlayCode) => {
 
 const MapOverlayBarComponent = ({
   currentMapOverlays,
-  isMeasureLoading,
   currentOrganisationUnitName,
-  onUpdateOverlayPeriod,
   mapOverlayHierarchy,
   defaultMapOverlay,
   currentMapOverlayCodes,
   onSetMapOverlay,
+  onSetDisplayedMapOverlays,
   onClearMeasure,
 }) => {
   const [hasNeverBeenChanged, setHasNeverBeenChanged] = useState(true);
-  const [pinnedOverlay, setPinnedOverlay] = useReducer(pinnedOverlayCodeReducer, null);
   const [maxSelectedOverlays, setMaxSelectedOverlays] = useState(1);
+  const [pinnedOverlay, setPinnedOverlay] = useReducer(pinnedOverlayCodeReducer, null);
   useEffect(() => {
     const { length } = currentMapOverlayCodes;
     if (length > maxSelectedOverlays) {
       setMaxSelectedOverlays(length);
     }
+    if (!currentMapOverlayCodes.includes(pinnedOverlay)) {
+      setPinnedOverlay(null);
+    }
   }, [currentMapOverlayCodes]);
+  useEffect(() => {
+    if (maxSelectedOverlays === 1) {
+      setPinnedOverlay(null);
+      onSetDisplayedMapOverlays(currentMapOverlayCodes);
+    }
+    if (maxSelectedOverlays < currentMapOverlayCodes.length) {
+      onSetMapOverlay(currentMapOverlayCodes.slice(0, maxSelectedOverlays));
+    }
+  }, [maxSelectedOverlays]);
```

Comment:
My understanding on `useEffect` is it usually do fetch or async function in it. 

One example is we do `fetch(A,B,C)` in useEffect, we want to put `[A,B,C]` as dependency because we want to fetch when one of them has been changed. But if we put [A,B] in the dependency array, when `A` is changed, fetch should use the latest `C` (if `C` has been updated). 

In our case `currentMapOverlayCodes` is more like the `C`. We don't want to run this useEffect if `currentMapOverlayCodes` has been changed. 

I've also tested it locally to make sure it is using the latest `currentMapOverlayCodes`.

Reply:
Oh I didn't know that we don't want to re-fetch when `currentMapOverlayCodes` changes. In that case, what you did seems perfect, thanks for clarifying @billli0 !